[multiple packages] Fix broken typings causing `check-packages` failures

bbarthec opened the pull request:
# Why

Several packages are constantly failing on CI on `check-packages` task.
I suspect that the recent typescript upgrade caused this, because most problems originate from `lib.dom.d.ts` shipped with `typescript` packages.

<details>
  <summary markdown="span">Expand to see the full list of issues (affected packages)</summary>

![et check-packages output](https://user-images.githubusercontent.com/16623003/138970180-cba1f4b8-9b7d-4fce-b36e-3cb3c28c2ef3.png)
</details>

# How

I've went package by packages and fixed the issues that were mainly connected with missing typings.
I've created several `./src/ts-definitions/lib.dom.d.ts` files with more specific type(s) to address the TS compilation problems.
In many packages there are parts that are very similar (or event exactly the same), but I've failed to create a single source of better typings. I've tried creating a separate `packages/expo-ts-definitions` package, but failed to make TS recognise it as a source of types (tried `typeRoots`, `types` and many other variations).

# Test Plan

`et check-packages -a` does not report any single error anymore.

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

bbarthec committed:
[av][Web] Fix missing/invalid typings

bbarthec committed:
[module-scripts] Add './ts-declarations' to sources by default

bbarthec committed:
[sensors][Web] Enhance custom typings for missing 'requestPermission' method

bbarthec committed:
[av] Refactor custom typings

bbarthec committed:
[module-scripts] Refactor changes

bbarthec committed:
[barcode-scanner][Web] Fix incorrect typings

bbarthec committed:
[camera][Web] Fix missing typings

bbarthec committed:
[cellular][Web] Fix missing typings

bbarthec committed:
[dev-client][plugin] Fix broken typings

bbarthec committed:
[font] Fix invalid type

bbarthec committed:
[gl] Add missing types by adding "@types/offscreencanvas" to dev dependencies

bbarthec committed:
[modules-autolinking] Fix broken type

bbarthec committed:
[modeles-core] Rebuild build directory

bbarthec committed:
[permissions][Web] Fix missing typings

bbarthec committed:
[updates][Web] Fix broken typings

bbarthec committed:
[gl] Fix tsconfig for tests

bbarthec committed:
[gl] Fix linter issue

bbarthec committed:
Update packages/expo-permissions/src/ExpoPermissions.web.ts

Co-authored-by: James Ide <ide@users.noreply.github.com>

bbarthec committed:
[permissions] Rebuild build directory

ide reviewed (approved):
Try making the font promises non-undefined (see my response in the comments) before merging.

tsapeta reviewed (approved):
I agree with James ☝️ 

Just to note, some of the issues were caused by this change in `useUnknownInCatchVariables` default value: https://devblogs.microsoft.com/typescript/announcing-typescript-4-4/#use-unknown-catch-variables
We can also consider turning it on, so that errors are treated as `unknown`

bbarthec committed:
[no ci][font] Remove undefined from the type definition and use 'hasOwnProperty' instead of direct bracket access in the if statement

Removed `undefined` from `const loadPromises: { [name: string]: Promise<void> | undefined } = {};` constant definiotion type
Replaced direct bracked access `if (loadPromises[fontFamily])` with more ts-friendly `if (loadPromises.hasOwnProperty(fontFamily))` check that would not trigger ts error upon not called function in `if` statement.

Simek commented:
@bbarthec The changes in this PR have broken the TypeDoc data extraction for `expo-cellular`:

<img width="1151" alt="Screenshot 2021-10-31 at 17 29 29" src="https://user-images.githubusercontent.com/719641/139593256-7227e11c-9f9c-458f-8140-f4be226aa8a8.png">
